src/lib/flow/sol-flow-js.c

```diff
@@ -0,0 +1,1422 @@
+{
+    const char *name, *type_name;
+    const struct sol_flow_packet_type *packet_type;
+    struct flow_js_port_in *port_type;
+    uint16_t array_len, i;
+
+    if (!duk_has_prop_string(duk_ctx, -1, "in"))
+        return true;
+
+    duk_get_prop_string(duk_ctx, -1, "in");
+
+    if (!duk_is_array(duk_ctx, -1)) {
+        SOL_ERR("'in' property of variable 'ports' should be an array.");
+        return false;
+    }
+
+    if (!duk_get_prop_string(duk_ctx, -1, "length")) {
+        SOL_ERR("Couldn't get 'in' length from 'ports' variable.");
+        return false;
+    }
+
+    array_len = duk_require_int(duk_ctx, -1);
+    duk_pop(duk_ctx); /* length value */
+
+    if (array_len == 0)
+        return true;
+
+    sol_vector_init(ports_in, sizeof(struct flow_js_port_in));
+
+    for (i = 0; i < array_len; i++) {
+        if (!duk_get_prop_index(duk_ctx, -1, i)) {
+            SOL_WRN("Couldn't get input port information from 'ports.in[%d]', ignoring this port creation...", i);
+            duk_pop(duk_ctx);
+            continue;
+        }
+
+        if (!duk_get_prop_string(duk_ctx, -1, "name")) {
+            SOL_WRN("Input port 'name' property is missing on 'ports.in[%d]', ignoring this port creation... "
+                "e.g. '{ name:'IN', type:'boolean' }'", i);
+            duk_pop_2(duk_ctx);
+            continue;
+        }
+
+        if (!duk_get_prop_string(duk_ctx, -2, "type")) {
+            SOL_WRN("Input port 'type' property is missing on 'ports.in[%d]', ignoring this port creation... "
+                "e.g. '{ name:'IN', type:'boolean' }'", i);
+            duk_pop_3(duk_ctx);
+            continue;
+        }
+
+        name = duk_require_string(duk_ctx, -2);
+        type_name = duk_require_string(duk_ctx, -1);
+
+        packet_type = get_packet_type(type_name);
+        if (!packet_type) {
+            SOL_WRN("Input port type '%s' is an invalid packet type on 'ports.in[%d]', ignoring this port creation...", type_name, i);
+            duk_pop_3(duk_ctx);
+            continue;
+        }
+
+        port_type = sol_vector_append(ports_in);
+        SOL_NULL_CHECK(port_type, false);
+
+        port_type->type.api_version = SOL_FLOW_PORT_TYPE_IN_API_VERSION;
+        port_type->type.packet_type = packet_type;
+        port_type->type.process = flow_js_port_process;
+        port_type->type.connect = flow_js_port_in_connect;
+        port_type->type.disconnect = flow_js_port_in_disconnect;
+
+        port_type->name = sol_arena_strdup(str_arena, name);
+        SOL_NULL_CHECK(port_type->name, false);
+
+        port_type->type_name = sol_arena_strdup(str_arena, type_name);
+        SOL_NULL_CHECK(port_type->type_name, false);
+
+        duk_pop_3(duk_ctx);
+    }
+
+    duk_pop(duk_ctx); /* in value */
+
+    return true;
+}
+
+static bool
+setup_ports_out(struct duk_context *duk_ctx, struct sol_arena *str_arena, struct sol_vector *ports_out)
```

Comment:
These two functions are very similar; can't they be just one function that's called twice?